### PR TITLE
microchip xec: Introduce I2c slave support

### PR DESCRIPTION
**Pull request by rgundi**
This patchset introduces i2c_slave support for microchip i2c driver

### CODE REVIEW

**Comment by scottwcpg** [comment]
I'm not sure this code is following section 5.1.6.1 Table 5-4 Re-enabling the Slave Receiver in the spec.  If the slave receiver is not being re-enabled after a STOP was issued then that could account for AAS not being set when the external master starts another transaction.  You could try writing 0xC1 to control per Table 5-4 in our wait_for_bus_idle routine just before it returns 0(success).

**Comment by rgundi** [comment]
@scottwcpg are you referring to functions wait_bus_free() and/or wait_completion()? These are called only when the i2c controller is a Master (polling mode). When the i2c controller is configured as a slave, almost everything happens under i2c_xec_bus_isr() in the interrupt context. Also, the configuration you are talking about is already a part of restart_slave() function which gets called whenever the i2c controller is configured as a slave.

**Comment by scottwcpg** [comment]
@rgundi I'm interpreting the MCHP spec. section 5.1.6.1 as required whenever a STOP is generated by the controller in Master mode or by an external controller issuing a STOP.  With respect to seeing data that's not there, if you read the I2C Data contains what ever is on the bus when clocks are present (generated by the controller in master mode) or from the external master.  If AAS is not set the I2C Data register will have something in it and it is not surprising the value is the address. The problem appears to be the controller is not ACK'ing its slave address.  This is why I believe section 5.1.6.1 in the spec. applies.  If the CONTROL register does not have the ACK bit (bit 0) set then it will not ACK a matching slave address.  Also, do you have the idle interrupt enabled (bit 29) in the CONFIG register? I don't recommend enabling the idle interrupt because the controller will fire the interrupt every time it's idle resulting in an interrupt when you may not want one.

File: drivers/i2c/i2c_mchp_xec.c
Comment (scottwcpg) [inline drivers/i2c/i2c_mchp_xec.c:499]: The HW can assert AAS first if an address match then PIN after the 9th clock. If you have AAS interrupt enabled you may enter the ISR with AAS==1 and PIN==1. For this case line 499 is false and the ISR runs the next if-else to write or read data.  Can you instrument the ISR to check if AAS==1 and PIN==1 at line 499 when the problem occurs?

**Review by scottwcpg — CHANGES_REQUESTED** [review]
Reading the line levels using the BB Control register does not work.  It always return both lines as High unless BB mode is enabled.  If BB mode is enabled then a HW MUX switches SDA/SCL away from the I2C logic to BB logic.  This causes I2C Status bits to change.  BB mode must be disabled to switch the lines back to I2C.  Code must check I2C Status and clean it.  Reading the pins using GPIO is safer but will require significant driver changes.

File: drivers/i2c/i2c_mchp_xec.c
Comment (franciscomunoz) [inline drivers/i2c/i2c_mchp_xec.c:535]: @rgundi Don't you plan to address the TODO comment in this PR?  Everything else looks good

Comment (rgundi) [reply drivers/i2c/i2c_mchp_xec.c:535]: Well.. An I2C slave should send a NACK to the master only if it is busy doing something for a long time (where mere clock stretching is not sufficient). It is definitely not needed for the application that i am pursuing right now. So, i don't plan to address it at least in this PR.

**Review by franciscomunoz — APPROVED** [review]

**Review by scottwcpg — DISMISSED** [review]

**Review by pabigot — CHANGES_REQUESTED** [review]
This seems to incorporate #28837; is that still being worked?  If it's superseded by this one, please close it.

Same change request here: the subject line of the first commit must specify that it applies to this driver, not all of I2C.

Also: Have you verified that tests/drivers/i2c/i2c_slave_api passes with this?  It would require two I2C controllers, and a board overlay in the test.

**Comment by rgundi** [comment]
@pabigot #28837 is very much valid. It's just that i wasn't sure how to push the rest of this patch for review while #28837 was under review. I thought that review would get approved faster and i could simply rebase this patch while i was addressing the review comments. I will take out this commit once #28837 gets merged.

I will modify the first commit making it explicit that it applies only to this driver.

I haven't verified this with tests/drivers/i2c/i2c_slave_api. But, it is part of an application (not open sourced) and i can confirm it is working there.

**Comment by pabigot** [comment]
> I haven't verified this with tests/drivers/i2c/i2c_slave_api. But, it is part of an application (not open sourced) and i can confirm it is working there.

That's fine, but unless this is a private SOC that nobody else can ever use, it should be verified with the in-tree test as well.

**Comment by rgundi** [comment]
> > I haven't verified this with tests/drivers/i2c/i2c_slave_api. But, it is part of an application (not open sourced) and i can confirm it is working there.
> 
> That's fine, but unless this is a private SOC that nobody else can ever use, it should be verified with the in-tree test as well.

Oh. This is definitely NOT a private SoC nobody is going to use. I can confirm the SoC contains multiple I2C ports. I read the README.txt and i see that SCL and SDA on one port need to be shorted to the corresponding ones on the other. Also, i guess i may need an overlay for the specific board i am using. May i know what else is required?

**Comment by pabigot** [comment]
> i see that SCL and SDA on one port need to be shorted to the corresponding ones on the other. Also, i guess i may need an overlay for the specific board i am using. May i know what else is required?

That should be all that's necessary.  Use the existing overlays as a model.

**Comment by rgundi** [comment]
#28837 is merged and i have rebased this. Also, have modified the commit messages as per suggestion by @pabigot.

**Review by pabigot — CHANGES_REQUESTED** [review]
AFAICT the changes in 947db1afc5 "drivers: i2c_mchp_xec: fix build errors when slave mode is disabled" are fixing problems introduced in earlier commits in the PR.  Please integrate the changes into those commits.

Also please confirm the test passes on the hardware (it's not one I can independently verify).

**Comment by rgundi** [comment]
> AFAICT the changes in 947db1afc5 "drivers: i2c_mchp_xec: fix build errors when slave mode is disabled" are fixing problems introduced in earlier commits in the PR. Please integrate the changes into those commits.
Done
> 
> Also please confirm the test passes on the hardware (it's not one I can independently verify).
This is Work in progress. Will keep you posted.

**Comment by pabigot** [comment]
> This is Work in progress. Will keep you posted.

OK.

Please also re-order the commits so the binding update for GIRQ flags is before the driver update that requires those properties.

**Comment by rgundi** [comment]
> Please also re-order the commits so the binding update for GIRQ flags is before the driver update that requires those properties.
Done. Also squashed relevant commits to have only 3 now.

**Comment by pabigot** [comment]
Looks reasonable, just waiting on confirmation the test passes https://github.com/zephyrproject-rtos/zephyr/pull/29475#issuecomment-753926363

**Comment by scottwcpg** [comment]
Test program fails.
*** Booting Zephyr OS build zephyr-v2.4.0-3006-g72998f07e6b1  ***
Running test suite test_eeprom_slave
===================================================================
START - test_eeprom_slave
I: Found EP0 EEPROM_0 on I2C Master device I2C0 at addr 54
I: Found EP1 EEPROM_1 on I2C Master device I2C1 at addr 56
I: EEPROM EEPROM_0 Attached !
I: EEPROM EEPROM_1 Attached !
I: Start full read. Master: I2C1, address: 0x54
E: Device is registered as slave

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/i2c/i2c_slave_api/src/main.c:60: run_full_read: (ret not equal to 0)
Failed to read EEPROM
 FAIL - test_eeprom_slave
===================================================================
Test suite test_eeprom_slave failed.
===================================================================
PROJECT EXECUTION FAILED
I looked at the registers for I2C contoller's 0 and 1 and the port is set for 4 and 5 respectively.  I thought the test used ports 0 and 1?  The boards/arm/mec1501_modular DTS file has port_sel 0 and 1 for controllers 0 and 1 respectively.

**Comment by scottwcpg** [comment]
More debug. In my EVB test setup I have a non-Zephyr firmware in SPI flash loaded by MEC152x Boot-ROM. This firmware (chromiumos EC code) uses I2C controller's 0 and 1.  If I download the Zephyr slave I2C test via JTAG and run it I observed the I2C controller's are not reset and re-configured.  Putting a break point on i2c_xec_configure shows it is never called.  The first I2C transaction the test does is:
device_get_binding
i2c_slave_driver_register - This calls the driver and the slave address is programmed but this API assumes the driver has previously configured the controller which is not true in this case.
i2c_burst_read - Calls driver i2c_write_read which sets up a call to driver i2c_transfer which assumes the driver has  previously configured the controller.
Are you assuming the I2C master tests are run first configuring the controller and then you run the I2C slave test without resetting the DUT?
Does the driver need logic to detect it has not been configured and force a configure call similar to the SPI drivers?

**Comment by rgundi** [comment]
@pabigot Could you please comment on Scott's findings?
@scottwcpg Are you running this on mec1501modular_assy6885 or did you make modifications to run this on an EVB?

**Comment by pabigot** [comment]
> @pabigot Could you please comment on Scott's findings?

The test overlay looks right, so I suspect Scott's hypothesis is correct that it's a driver bug.

**Comment by rgundi** [comment]
> > @pabigot Could you please comment on Scott's findings?
> 
> The test overlay looks right, so I suspect Scott's hypothesis is correct that it's a driver bug.

I had a relook at this driver as well as several other i2c drivers in the repo. Some noteworthy points below.
1) Most i2c drivers have the bus speed hardcoded at build time (via DTS). Some i2c drivers like mchp, esp32, npc etc have it configurable in runtime. There seems to be no uniformity in this.
2) Currently only i2c_ll_stm32.c supports i2c slave. I see that i2c_stm32_init has a call to i2c_stm32_runtime_configure. Not sure if that’s the right thing to do as some i2c drivers (refer point 1 above) may have configurable i2c frequency values.
3) In light of point 2, we can add a line to i2c_xec_init() invoking the function i2c_xec_configure() with speed set to I2C_SPEED_STANDARD. It should be alright since this is something that is already being done in i2c_nios2.c.

@pabigot @scottwcpg your comments please.

**Comment by pabigot** [comment]
> 1. Most i2c drivers have the bus speed hardcoded at build time (via DTS). Some i2c drivers like mchp, esp32, npc etc have it configurable in runtime. There seems to be no uniformity in this.

It's probably due to following Linux where this is a controller property that can be defaulted from devicetree, but changed at runtime if necessary.

> 2. Currently only i2c_ll_stm32.c supports i2c slave. I see that i2c_stm32_init has a call to i2c_stm32_runtime_configure. Not sure if that’s the right thing to do as some i2c drivers (refer point 1 above) may have configurable i2c frequency values.

I suspect it's justifiable given 1.

> 3. In light of point 2, we can add a line to i2c_xec_init() invoking the function i2c_xec_configure() with speed set to I2C_SPEED_STANDARD. It should be alright since this is something that is already being done in i2c_nios2.c.

Yes, please try that.

**Comment by rgundi** [comment]
> > 1. In light of point 2, we can add a line to i2c_xec_init() invoking the function i2c_xec_configure() with speed set to I2C_SPEED_STANDARD. It should be alright since this is something that is already being done in i2c_nios2.c.
> 
> Yes, please try that.

Done. @scottwcpg Please run the test with the updated changes.

**Comment by scottwcpg** [comment]
Tried the latest code.  Test fails on first call to driver from test function run_full_speed calling i2c_burst_read.
The call into driver i2c_xec_transfer return -EBUSY from the if (data->slave_attached) statement line 434. Test code is trying to read I2C address 0x54 which is the registered slave address of I2C controller 0.  What is interesting is the device pointer appears to be correct: name="I2C1".  The if logic is checking struct i2c_slave_config.slave_attached flag.  I thought it should check if the addr parameter passed to i2c_xec_transfer matches one of the two slave addresses programmed into the current controller.  The test program is passing address 0x54 and I2C1 has slave address 0x56 so it's OK for I2C1 to communicate with 0x54.  NOTE: These I2C controllers can have two slave addresses (bits[7:0] and bits[15:8]) of the controller own address register.  The driver is currently only using one slave address (bits[7:0]).

**Comment by rgundi** [comment]
> Tried the latest code. Test fails on first call to driver from test function run_full_speed calling i2c_burst_read.
> The call into driver i2c_xec_transfer return -EBUSY from the if (data->slave_attached) statement line 434. Test code is trying to read I2C address 0x54 which is the registered slave address of I2C controller 0. What is interesting is the device pointer appears to be correct: name="I2C1". The if logic is checking struct i2c_slave_config.slave_attached flag. I thought it should check if the addr parameter passed to i2c_xec_transfer matches one of the two slave addresses programmed into the current controller. The test program is passing address 0x54 and I2C1 has slave address 0x56 so it's OK for I2C1 to communicate with 0x54. NOTE: These I2C controllers can have two slave addresses (bits[7:0] and bits[15:8]) of the controller own address register. The driver is currently only using one slave address (bits[7:0]).

The intention of "if (data->slave_attached)" is to ensure i2c_xec_transfer() is never executed while it is configured as a slave. The design that we have right now will simply not work for this test. It currently doesn't support simultaneous master and slave.

**Comment by pabigot** [comment]
> The design that we have right now will simply not work for this test. It currently doesn't support simultaneous master and slave.

Even on two separate controllers?

**Comment by pabigot** [comment]
I guess if the MEC device can't support the test the options are to fix the test, add a separate board-specific test, or merge this without the ability to verify it continues to work as Zephyr evolves.  The last option doesn't seem like a good plan.  I would think fixing the test and/or driver so they work together should be possible, but without access to hardware I can't say how to make it happen.

**Comment by rgundi** [comment]
> > The design that we have right now will simply not work for this test. It currently doesn't support simultaneous master and slave.
> 
> Even on two separate controllers?

Yes. The problem is i2c_slave_driver_register() is being called for both the controllers right at the beginning of the test. This is putting both the controllers into slave mode. So, the xec driver cannot function as a master. It will be interesting to see what happens if the test is modified to call i2c_slave_driver_register() only on the controller which needs to be the slave.

**Comment by scottwcpg** [comment]
I hacked the test to register/unregister the follower based upon test flow.  Attached archive containing the test main.c and UART capture from my MEC152x EVB.  NOTE: I'm building using board mec1501modular_assy6885 since that's what the test is configured for.  The EVB has jumpers to connect I2C0 and I2C1 pins.
[xec_i2c_slave_api_test_mod_results.zip](https://github.com/zephyrproject-rtos/zephyr/files/5843458/xec_i2c_slave_api_test_mod_results.zip)

**Comment by pabigot** [comment]
Since the test incorrectly assumes the controllers support both roles without configuration, I'm modifying it to have that be controllable.  I'll have a PR or commit for that shortly for testing.

**Comment by pabigot** [comment]
@scottwcpg Could you please grab #31458 and see whether the driver passes now?  Dual-mode is not on by default, it's only enabled for automated testing on ST devices.  Thanks.

**Comment by scottwcpg** [comment]
Hi @pabigot 
Thanks! It works.
I applied PR31458 on top of PR29475.  Attached a zip of the UART capture.
[uart1_cap01.zip](https://github.com/zephyrproject-rtos/zephyr/files/5843971/uart1_cap01.zip)

**Comment by rgundi** [comment]
Thanks @pabigot and @scottwcpg . This can hopefully be merged now. Right in time before the window closes. @franciscomunoz @albertofloyd @nashif, please review.

**Review by pabigot — APPROVED** [review]

**Review by scottwcpg — APPROVED** [review]

**Pull request merged**
